### PR TITLE
Remove abstract sched

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
This follows the idea given in https://github.com/ankidroid/Anki-Android/issues/6819#issuecomment-669368458 to follow upstream and entirely stop using AbstractSched.

The main goal is that, when I ask android studio to show me the definition of a function, I see actual code

### CODE REVIEW

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.java [hunk 1]
```diff
@@ -169,7 +169,7 @@ public void ensureDeckTree() {
             addDeck(deckName);
         }
         getCol().getSched().deckDueTree();
-        AbstractSched sched = getCol().getSched();
+        SchedV2 sched = getCol().getSched();
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.java:172]: Aren't these Sched, and not SchedV2?

In places like this, AbstractSched makes sense - as we don't care about the Scheduler version

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedV2Test.java [hunk 2]
```diff
@@ -884,18 +884,18 @@ public void test_bury() throws Exception {
         col.reset();
         assertNull(col.getSched().getCard());
 
-        col.getSched().unburyCardsForDeck(AbstractSched.UnburyType.MANUAL);
+        col.getSched().unburyCardsForDeck(SchedV2.UnburyType.MANUAL);
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedV2Test.java:887]: Might as well remove the SchedV2 here

**Comment by mikehardy** [comment]
I see the point of having AbstractSched but this is something we are waffling on philosophically, while Arthur is trying to propose merge-able PRs. Can we slow down for a bit? (both on the speed of back and forth, and speed of refactoring PRs)?

The clearest way I can say this is that I have a strong preference for following upstream. Upstream doesn't have an AbstractSched object. So I want a good justification to have one here.

As I started with - I see the point of maybe having one so that "SchedV1" is not (technically) a V2 type, but this is getting close to ivory tower academic object oriented design to me. In practice, I want to refer to the python code and the java code side by side for as long as we are maintaining them, and because of language differences that gives us some Java warts. I'm not fussed by that

So, does it really matter if v1 will answer "instanceof SchedV2" with a true? is that why we will have maintenance work in the future? I'm skeptical

**Comment by Arthur-Milchior** [comment]
> The fact that `Sched` inherits from `SchedV2` complicates matters, as you have a variable of type `SchedV2` which is not a V2 Schedler. `AbstractSched` makes this distinction much more clear

That's more or less what I stated myself. But it seems @mikehardy disagreed in https://github.com/ankidroid/Anki-Android/issues/6819#issuecomment-669368458  . One of his argument was that if someone does not know the scheduler enough to understand that the type SchedV2 represents also sched V1 (As I wrote in a comment) then they should not deal with the scheduler. I am not convinced, but it makes sens.

Whatever decision we make, I must at least state that I really wish that the code change. Because we currently have a lot of code duplication, it's hard to know when the two scheduler are actually different and what is their difference. And when I ask to go to the definition of a method when I debug the code, I want to see the actual implementation. Since all scheduler is abtract, I never see any implementation and I loose a lot of time looking for the real code.

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedTest.java:172]: It is indeed a sched. However, the typing system has no way of knowing it. We can of course cast to Sched every time if we want. At worst, it makes test fails, and that would be an important indication

File: AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedV2Test.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/test/java/com/ichi2/libanki/sched/SchedV2Test.java:887]: Done

**Comment by Arthur-Milchior** [comment]
I realized it was too big. I splitted this in multiple smaller PR simpler to discuss. I hope some of them will be really simple to consider

**Comment by dae** [comment]
For what it's worth, the original plan was to migrate to v2 and drop support for v1 in a fairly short timeframe, so separate files seemed like the easiest approach at the time. I didn't expect the transition to 2.1 to drag out for as long as it did, and the separate files proved to be a giant maintenance burden :-(

sched inheriting from sched2 is a nod to the end goal of getting back to one scheduler, and since 2.0 syncing is now broken, we may finally be getting close to the point where the old scheduler can be retired.

**Comment by Arthur-Milchior** [comment]
Thank you for this bit of explanation. 
I strongly suspect that it's a point in favor of this PR and in defavor of the previous version. Actually using SchedV2 everywhere seems to make even more sens now. This probably means that I'll close all PRs which tried to add the code in common in AbstractSched and leave in SchedV2 only the code unique to scheduler V2

**Pull request closed without merging**